### PR TITLE
doc: add forecasting controller proposal

### PR DESCRIPTION
**Pull request by zwzhang0107**
### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+### Profile Resource Usage for FaaS Tasks
+A FaaS service is consist of functions, which receive requests and invoke the executor. Resource usage metrics are exposed
+for `Prometheus` as exporter.
+```
+function_cpu_usage{service="service-word-count", function="f-map-name", slice="xxx"} 1.2
+function_memory_usage{service="service-word-count", function="f-map-name", slice="xxx"} 1024000
+``` 
+We use the forecasting framework to aggregate the metric of each function, and predict the resource usage for right-sizing
+and load aware scheduling.
+
```
Comment (hormes) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:264]: This is a big topic, please add a section here describing our Milestone (aka our implementation path)

File: docs/images/forecasting/metric-apis.svg
Comment (saintube) [inline docs/images/forecasting/metric-apis.svg:1]: metrics.k8s.io

File: docs/images/forecasting/forecast-closure.svg
Comment (saintube) [inline docs/images/forecasting/forecast-closure.svg:1]: some types in figures `forecast-closure`, `forecasting-controller`, `metric-apis`
e.g. `Prometehus`

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+type DistributionValues struct {
+	Mean              resource.Quantity                `json:"mean,omitempty"`
+	Quantiles         map[string]resource.Quantity     `json:"quantiles,omitempty"`
+	StdDev            resource.Quantity                `json:"stddev,omitempty"`
+	TotalSamplesCount int                              `json:"totalSamplesCount,omitempty"`
+}
+```
+
+### How to assemble forecasting models to support multiple scenarios
+All forecasting stages designed as tool packages, which can be extended with various implementation (e.g. Prometheus
```
Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:232]: implementation -> implementations

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+	TotalSamplesCount int                              `json:"totalSamplesCount,omitempty"`
+}
+```
+
+### How to assemble forecasting models to support multiple scenarios
+All forecasting stages designed as tool packages, which can be extended with various implementation (e.g. Prometheus 
+or metric-server type `Collector`). We assemble all stages into a `Closure` for each target according to the spec, 
+then a module called `Forecast Runner` will run these `Closure` parallel.
+    
+We define a `Closure` for each target, include the metric and prediction model to use. `Closure` can be run independently
```
Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:236]: include -> including

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+by `Forecast Runner`.
+
+![forecast-closure](../images/forecasting/forecast-closure.svg)
+
+The top-level are controllers for different scenarios such as resource recommend and interference detection. We define
```
Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:241]: recommend -> recommendations

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+CRD for each controller and then the CR will be convert to a `Closure` we defined in forecast framework. The
```
Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:242]: be convert -> be converted

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+then a module called `Forecast Runner` will run these `Closure` parallel.
+    
+We define a `Closure` for each target, include the metric and prediction model to use. `Closure` can be run independently
+by `Forecast Runner`.
+
+![forecast-closure](../images/forecasting/forecast-closure.svg)
+
+The top-level are controllers for different scenarios such as resource recommend and interference detection. We define
+CRD for each controller and then the CR will be convert to a `Closure` we defined in forecast framework. The 
+`Forcast Runner` executes all `Closures` and produces the result through observer, which will be convert and write back
```
Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:243]: be convert -> be converted

Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:243]: write -> written

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+to CRD status field.
+
+![forecasting-controller](../images/forecasting/forecasting-controller.svg)
+
+## Examples
+
+### Detect Outliers for Container CPI Metric
+Container [CPI metric](https://static.googleusercontent.com/media/research.google.com/zh-CN//pubs/archive/40737.pdf) shows
+the runtime quality of applications, which keeps relative stable over serving time. We build a prediction model based on
```
Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:252]: relative -> relatively

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+historical metric data to determine a container disrupted by others if metric has outliers at present.
```
Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:253]: to determine whether a container is disrupted by others if the metrics has outliers at present

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+
+## Examples
+
+### Detect Outliers for Container CPI Metric
+Container [CPI metric](https://static.googleusercontent.com/media/research.google.com/zh-CN//pubs/archive/40737.pdf) shows
+the runtime quality of applications, which keeps relative stable over serving time. We build a prediction model based on
+historical metric data to determine a container disrupted by others if metric has outliers at present. 
+
+### Profile Resource Usage for FaaS Tasks
+A FaaS service is consist of functions, which receive requests and invoke the executor. Resource usage metrics are exposed
```
Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:256]: is consist -> consists

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+for `Prometheus` as exporter.
+```
+function_cpu_usage{service="service-word-count", function="f-map-name", slice="xxx"} 1.2
+function_memory_usage{service="service-word-count", function="f-map-name", slice="xxx"} 1024000
+``` 
+We use the forecasting framework to aggregate the metric of each function, and predict the resource usage for right-sizing
+and load aware scheduling.
```
Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:263]: load aware -> load-aware

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+
+## Alternatives
+Actually, we plan to implement each forecasting or detection model as an independent controller, and each controller
+finishes all work by its own (e.g. resource-recommendation, interference-detection). However, after taking a deep look
```
Comment (jasonliu747) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:267]: on its own

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+### Goals
+- Design a holistic framework with modules of profile control, forecasting and data scraping to support data prediction and detection.
+
+- Clarify the data type and transfer links, and how to save the storage and collection overhead.
+
+- Collaboration and division of builtin and self-extended metric exporters, and adaption of multiple using scenarios.
+
+### Non-Goals
+- Design of forecasting algorithm. The interference detection and resource prediction may share same prediction models,
```
Comment (LeoLiuYan) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:81]: If the framework does not include forecasting algorithm,  how do we guarantee the framework output accurate data?
@zwzhang0107

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+however, this is over-detailed now. 
+
+- Supports of MPA framework proposed in Kubernetes autoscaler. Although out work can run as a recommender for MPA after
+some protocol conversion works, we can discuss in following proposals.
+
+- Long-term optimization, even though we have already mentioned some ideas about saving transfer cost and storage overhead.
```
Comment (LeoLiuYan) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:87]: May be the storage implement of metrics has a variety of options? For example, InfluxDB. Default is `in memory`.

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+The optimization above relies on the analysis of historical metric data for prediction and detection. So we propose a general 
+forecasting framework, retrieving metric data and build models such as resource usage and runtime quality. These models
+can be used on Interference Detection for higher QoS, and Resource Prediction to improve resource efficiency. 
+
+## Motivation
+
+Several topics are still waiting to be proposed in Koordinator yet!
+
+- Interference Detection: Koordlet has provided several QoS Strategies to mitigate the noisy neighbor problem: setting different resource limit
+for pods for different QoS Class. These are pre-active and preventive measures, trying to prevent noisy neighbor happens.
```
[the remark this answers is not in the document]
Comment (zwzhang0107) [reply docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:56]: it is pre-active

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+
+These topics are all related to data analysis, so we need a general framework for prediction and detection to support 
+different scenarios. 
+
+### Goals
+- Design a holistic framework with modules of profile control, forecasting and data scraping to support data prediction and detection.
+
+- Clarify the data type and transfer links, and how to save the storage and collection overhead.
+
+- Collaboration and division of builtin and self-extended metric exporters, and adaption of multiple using scenarios.
```
[the remark this answers is not in the document]
Comment (zwzhang0107) [reply docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:78]: also have using scenarios.

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (zwzhang0107) [reply docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:81]: the xxx or xxxs both ok

Comment (zwzhang0107) [reply docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:81]: The `forecast-controller` definitely has prediction or detection algorithms, but this proposal does not include the detailed design of the implement of algorithms.

The algorithm should be implemented in the `Model` part, and we will show the code in later PRs or proposals.

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md (hunk 13 above)
Comment (zwzhang0107) [reply docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:87]: Yes It is, and especially for some algorithms depends on offline training. At the beginning, saving in memory is enough for some plain models.

File: docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+`Prometheus` is another widely used metric service, which also provides good extendability. User can collect metrics in their
+own components(e.g. daemon-set, operator), and expose as standard exporter format. Then these metrics can be accessed from
+`Prometheus` after the service monitor defined.
+
+Besides, Kubernetes also provides customized API as CRD and aggregate layer in `apiserver`, which can be used for metric
+report. Furthermore, we can define dedicated protocols as an internal metric service.
+
+`Prometheus` seems to be a good option before we take a deep look at the pros and cons.
+
+Pros:
```
Comment (ZiMengSheng) [inline docs/proposals/forecasting/20230613-framework-of-prediction-and-detection.md:124]: Promtheus Metrics can be adapted to Kubernetes Metric API by Prometheus Adapter. Thus we don't need a metric server.

**Review by ZiMengSheng — COMMENTED** [review]
/lgtm

**Review by hormes — APPROVED** [review]

**Pull request merged**